webhook/webhook_test.go

```diff
@@ -0,0 +1,189 @@
+package webhook
+
+import (
+	"testing"
+	"net/http"
+	"net/http/httptest"
+	"log"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"bytes"
+	"io/ioutil"
+
+	"github.com/stretchr/testify/suite"
+	"github.com/gophish/gophish/webhook"
+	"github.com/stretchr/testify/assert"
+)
+
+type WebhookSuite struct {
+	suite.Suite
+}
+
+type mockSender struct {
+	client *http.Client
+}
+
+func newMockSender() *mockSender {
+	ms := &mockSender {
+		client: &http.Client{},
+	}
+	return ms
+}
+
+func (ms mockSender) Send(endPoint webhook.EndPoint, data interface{}) error {
+	log.Println("[test] mocked 'Send' function")
+	return nil
+}
+
+func (s *WebhookSuite) TestSendMocked() {
+	mcSnd := newMockSender()
+	endp1 := webhook.EndPoint{URL: "http://example.com/a1", Secret: "s1"}
+	d1 := map[string]string {
+		"a1": "a11",
+		"a2": "a22",
+		"a3": "a33",
+	}
+	err := mcSnd.Send(endp1, d1)
+	s.Nil(err)
+}
+
+func (s *WebhookSuite) TestSendReal() {
+	expectedSign := "4775314ed81be378b2b14f18ac29a6db0eb83b44ed464a000400d43100c8a01e"
+	successfulHttpResponseCode := 200
+	secret := "secret456"
+	d1 := map[string]interface{} {
+		"key11": "val1",
+		"key22": "val22",
+		"key33": map[string]string {
+			"key4": "val444",
+		},
+	}
+
+	hClient := &http.Client{}
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Println("[test] running the server...")
+
+		realSign := r.Header.Get(webhook.SignatureHeader)
+		assert.Equal(s.T(), expectedSign, realSign)
+
+		neHeader := r.Header.Get("non-existing-header")
+		assert.Equal(s.T(), neHeader, "")
+
+		contTypeJsonHeader := r.Header.Get("Content-Type")
+		assert.Equal(s.T(), contTypeJsonHeader, "application/json")
+
+		body, err := ioutil.ReadAll(r.Body)
+		s.Nil(err)
+
+		var d2 map[string]interface{}
+		err = json.Unmarshal(body, &d2)
+		s.Nil(err)
+
+		key1RealVal := d2["key1"]
```

Comment:
I'm not familiar enough with the details of `assert` to know if you could just do `assert.Equal(s.T(), d1, d2` but it might be worth a shot.

Reply:
Actually, I have an error that there's no access to the "sign" method; looking into it.

Reply:
fixed